File: 1507-check-if-there-is-a-valid-path-in-a-grid/check-if-there-is-a-valid-path-in-a-grid.py
class Solution:
    def hasValidPath(self, grid: List[List[int]]) -> bool:
        m,n = len(grid), len(grid[0])

        dirs = {
            1: [(0,-1),(0,1)], #LR
            2: [(-1,0),(1,0)], #UD
            3: [(0,-1),(1,0)],#LD
            4: [(0,1),(1,0)],  #RD
            5: [(0,-1),(-1,0)], #LU
            6: [(0,1),(-1,0)],  #RU
        }
        visited = [[False]*n for _ in range(m)]

        def dfs(i, j):
            visited[i][j] = True
            if (i==m-1 and j==n-1):
                return True
            
            val = grid[i][j]
            for dir in dirs[val]:
                x = i+dir[0]
                y = j+dir[1]

                flag = False
                if (0<=x<m and 0<=y<n):
                    new_val = grid[x][y]
                    for new_dir in dirs[new_val]:
                        if (x+new_dir[0]==i and y+new_dir[1]==j):
                            flag = True
                            break
                if (not flag): 
                    continue
                if (not visited[x][y]):
                    if (dfs(x,y)):
                        return True
            
            return False

        return dfs(0,0)

"""
Key:
    current cell can go to next cell
    AND
    next cell can go back to current cell

"""

File: 1507-check-if-there-is-a-valid-path-in-a-grid/check-if-there-is-a-valid-path-in-a-grid.cpp
class Solution {
public:
    vector<vector<pair<int,int>>> dirs={
        {{{0,0},{0,0}}},
        {{{0,-1},{0,1}}},//1
        {{{1,0},{-1,0}}},//2
        {{{0,-1},{1,0}}},//3
        {{{0,1},{1,0}}},//4
        {{{0,-1},{-1,0}}},//5
        {{{0,1},{-1,0}}},//6
    };

    bool dfs(int i, int j, vector<vector<int>>&grid, vector<vector<bool>>& visited){
        int m = grid.size();
        int n = grid[0].size();

        visited[i][j]=true;
        if (i==m-1 && j ==n-1)
            return true;
        int val=grid[i][j];

        for(int k=0; k<dirs[val].size();++k){
            int i_ = i+dirs[val][k].first;
            int j_ = j+dirs[val][k].second;

            bool flag = false;

            if (i_>=0 && i_<m && j_>=0 && j_<n){
                int new_val = grid[i_][j_];
                for(auto&[x,y]:dirs[new_val]){
                    if (i_+x == i && j_+y==j){
                        flag = true;
                        break;
                    }
                }
                if (flag == false){
                    continue;
                }
                if (visited[i_][j_] == false){
                    if (dfs(i_,j_, grid, visited)){
                        return true;
                    }
                }
            }  
        } 
        return false;
    }
    bool hasValidPath(vector<vector<int>>& grid) {
        int m = grid.size(), n=grid[0].size();
        vector<vector<bool>> visited(m,vector<bool>(n,false));
        return dfs(0,0,grid,visited);
    }
};